### ISSUE TITLE
Regular "Not connected" errors on `linkerd-proxy` on DigitalOcean Kubernetes

### ISSUE THREAD

**Issue by nickjackson** [linkerd/linkerd2#3735]
### What is the issue?
Installing linkerd2 manually or via the marketplace app on Digital Ocean's managed Kubernetes offering causes `linkerd-proxy` to spout the following errors roughly every minute.

```
admin={server=admin listen=0.0.0.0:4191 remote=10.244.1.39:42744} linkerd2_proxy::app::admin::serve_http error serving admin: Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
```

The remote IP, in this case, is what I believe to be the cilium `Internal-Node IPv4` aka `v4CiliumHostIP.IPv4` which I found in some cilium agent logs.

### How can it be reproduced?
1. Spin up a Digital Ocean Kubernetes Cluster
2. Install linkerd2
3. Look through the logs of any `linkerd-proxy` ( I seem to have luck with the `linkerd-prometheus` proxy) and the error should appear

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ no invalid service profiles

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

Status check results are √
```

### Environment

- Kubernetes Version: v1.16.2
- Cluster Environment:  DigitalOcean (1.16.2-do.0)
- Linkerd version: 2.6

### Possible solution
I'm not sure of the cause to suggest a fix. An uneducated guess might be that DigitalOcean is trying to scrape metrics from the proxy from outside the mesh?

### Additional context
I apologise if this is not the right place to put this issue, I needed somewhere to send DigitalOcean Support and keep everyone on the same page.

Linkerd and the internal apps seem unaffected by this.

**Comment by tlipinski** [comment]
I was about to report the same issue, however it most likely has nothing to do with DigitalOcean because I'm able to reproduce it on Hetzner cloud (non-managed Kubernetes):
```
 linkerd-proxy ERR! [   723.762493s] admin={server=admin listen=0.0.0.0:4191 remote=10.244. │
│ 2.1:53458} linkerd2_proxy::app::admin::serve_http error serving admin: Error(Shutdown, Os  │
│ { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })  
```
A few observations from me:
- it reports the error on Kubernetes 1.16.1 and 1.16.3
- it **does not** report such error on Kubernetes 1.15.5
- I see this error in my services' `linkerd-proxy` instances but only after they receive some traffic
- when I exec shell on my service and curl to `http://localhost:4191/metrics` then I get some metrics from `linkerd-proxy`.
- even though linkerd-proxy reports the error, services work as expected including communication between them

I can provide any other details necessary for investigation.

**Comment by nickjackson** [comment]
Hi @tlipinski,

I just have a couple questions 

1. Are you able to identify what that remote IP is? (10.244.2.1)

2. Do you have any firewall rules or network policy in place that might be getting in the way?

Thanks,
Nick

**Comment by tlipinski** [comment]
@nickjackson 
10.244.2.1 is an address of cni0 network which I believe was created by flannel. Cluster was created with recommended `--pod-network-cidr=10.244.0.0/16`. Apart from that I tested it on a very basic cluster created from scratch:
- single master node
- flannel `quay.io/coreos/flannel:v0.11.0-amd64`
- linkerd 2.6
- a few toy services which call each other for testing purposes
No network policies, any custom iptables rules etc.
Somehow on Kubetnetes 1.15.5 `linkerd-proxy` it didn't report any errors so maybe 1.16 introduced some breaking changes.

**Comment by ihcsim** [comment]
@nickjackson @tlipinski Thanks for the info.

> it reports the error on Kubernetes 1.16.1 and 1.16.3
> it does not report such error on Kubernetes 1.15.5

This piece of info is very useful. To date, I don't think we have done much extensive testing on K8s 1.16+. 

> Linkerd and the internal apps seem unaffected by this.

Good to know.

**Comment by tlipinski** [comment]
@nickjackson did you try running Linkerd 2.7? In my case it no longer reports "Not connected" errors on the same cluster as last time - Kubernetes 1.16.3. I haven't tried 1.17 yet.

**Comment by olix0r** [comment]
Newer versions of Linkerd (e.g., `edge-20.3.4`) have been updated to handle service discovery differently. If you're still experiencing these issues, I recommend annotating your workload with `config.linkerd.io/proxy-version: edge-20.3.4`. If you test this, please report back! These changes will be released soon in `stable-2.7.1`.
